plugin/framework: Add Resource Import page (#2032)
hashicorp/terraform-website: content/plugin/framework/resources/import.mdx

---
page_title: 'Plugin Development - Framework: Resource Import'
description: >-
  How to support resource import using the provider development framework.
---

# Resource Import

Terraform allows practitioners to bring existing infrastructure under management using the [`terraform import` command](/cli/commands/import). Resources implement the `ImportState` method, which must specify enough Terraform state to allow the [`tfsdk.Resource`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#Resource) to be successfully refreshed via the `Read` method or return an error.

## Single Attribute

The [`tfsdk.ResourceImportStatePassthroughID` function](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ResourceImportStatePassthroughID) is available to write the final argument (import identifier) to the `terraform import` command to an attribute.

> For example, if the `tfsdk.Resource` implementation has the following `ImportState` method:

Reviewer: ```suggestion
In the following example, the `terraform import` command passes the import identifier to the `id` attribute in Terraform state.
```

```go
func (r exampleResource) ImportState(ctx context.Context, req tfsdk.ImportResourceStateRequest, resp *tfsdk.ImportResourceStateResponse) {
    tfsdk.ResourceImportStatePassthroughID(ctx, tftypes.NewAttributePath().WithAttributeName("id"), req, resp)
}
```

> The `id` attribute in the Terraform state will receive the import identifier from the `terraform import` command.

Reviewer: I just like to try to tell folks what the example is showing before we give it to them. 

```suggestion
```

> 

Author: Just to leave one whitespace line after pulling in the other suggestion which deletes a line.

```suggestion
```

## Multiple Attributes

> Resources that require setting multiple attributes to successfully refresh in the `Read` method will require custom logic in the `ImportState` method. Use the import identifier from the [`tfsdk.ImportResourceStateRequest`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ImportResourceStateRequest), perform the custom logic, then [set state data](/plugin/framework/writing-state) in the [`tfsdk.ImportResourceStateResponse`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ImportResourceStateResponse).

Reviewer: ```suggestion
When the `Read` method requires multiple attributes to refresh, you must write custom logic in the `ImportState` method. Specifically, the implementation must:

1. Use the import identifier from the [`tfsdk.ImportResourceStateRequest`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ImportResourceStateRequest). 
1. Perform the custom logic.
1. [Set state data](/plugin/framework/writing-state) in the [`tfsdk.ImportResourceStateResponse`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ImportResourceStateResponse).
```

    idParts := strings.Split(req.ID, ",")

    if len(idParts) != 2 || idParts[0] == "" || idParts[1] == "" {
        resp.Diagnostics.AddError(
            "Unexpected Import Identifier",
            fmt.Sprintf("Expected import identifier with format: attr_one,attr_two. Got: %q", req.ID),
        )
        return
    }

    resp.Diagnostics.Append(resp.State.SetAttribute(ctx, tftypes.NewAttributePath().WithAttributeName("attr_one"), idParts[0])...)
    resp.Diagnostics.Append(resp.State.SetAttribute(ctx, tftypes.NewAttributePath().WithAttributeName("attr_two"), idParts[1])...)
}
```

## Not Implemented

The [`tfsdk.ResourceImportStateNotImplemented` function](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ResourceImportStateNotImplemented) is available to return an error to practitioners if the resource does not implement import support. For example:

```go
func (r exampleResource) ImportState(ctx context.Context, req tfsdk.ImportResourceStateRequest, resp *tfsdk.ImportResourceStateResponse) {
    tfsdk.ResourceImportStateNotImplemented(ctx, "", resp)
}
```

> When a practitioner attempts to run the `terraform import` command against a resource with this implementation, Terraform CLI will return this default messaging:

Reviewer: ```suggestion
When a practitioner runs `terraform import`, Terraform CLI will return:
```

```console
$ terraform import example_resource.example some-identifier
example_resource.example: Importing from ID "some-identifier"...
╷
│ Error: Resource Import Not Implemented
│ 
│ This resource does not support import. Please contact the provider developer for additional information.
╵
```

> Modifying the details parameter of the function allows for customized recommendations, such as other actions the practitioner can take.

Author: I can definitely see the potential for confusion here! The function signature names the parameter `details` to align with the concept that it is changing the "details" section of the error diagnostic it is creating. I wonder if it might be slightly clearer to further mention replacing the `""` in the above example.

> Modifying the details parameter of the function allows for customized recommendations, such as other actions the practitioner can take.

Reviewer: FWIW - I like the idea you just had :-)

        )
        return
    }

    resp.Diagnostics.Append(resp.State.SetAttribute(ctx, tftypes.NewAttributePath().WithAttributeName("attr_one"), idParts[0])...)
    resp.Diagnostics.Append(resp.State.SetAttribute(ctx, tftypes.NewAttributePath().WithAttributeName("attr_two"), idParts[1])...)
}
```

## Not Implemented

If the resource does not support `terraform import`, use the [`tfsdk.ResourceImportStateNotImplemented` function](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ResourceImportStateNotImplemented) to return an error. The following example returns a default error message:

```go
func (r exampleResource) ImportState(ctx context.Context, req tfsdk.ImportResourceStateRequest, resp *tfsdk.ImportResourceStateResponse) {
    tfsdk.ResourceImportStateNotImplemented(ctx, "", resp)
}
```

When a practitioner runs `terraform import`, Terraform CLI will return:

```console
$ terraform import example_resource.example some-identifier
example_resource.example: Importing from ID "some-identifier"...
╷
│ Error: Resource Import Not Implemented
│ 
│ This resource does not support import. Please contact the provider developer for additional information.
╵
```

> You can also modify the `details` parameter of the function to include customized recommendations, such as other actions the practitioner can take.

Author: ``````suggestion
You can also modify the `details` parameter of the function to include customized recommendations, such as other actions the practitioner can take. For example:

```go
func (r exampleResource) ImportState(ctx context.Context, req tfsdk.ImportResourceStateRequest, resp *tfsdk.ImportResourceStateResponse) {
    tfsdk.ResourceImportStateNotImplemented(ctx, "Custom error message.", resp)
}
```
``````
